File: wakakusa/demo/counter.rb
# The reference: everything in this file is what Wakakusa takes. The
# app is an object, its state is its instance variables, and a handler
# is a block that closes over it.
#
#   wakakusa run  demo/counter.rb
#   wakakusa gate demo/counter.rb --script "click:+1,input:Momo"
require "wakakusa"

class Counter
  def initialize
    @count = 0
    @name = ""
  end

  def view
    column(
      text("count: #{@count}", size: 34.0),
      row(
        button("+1") { @count += 1 },
        button("+10") { @count += 10 },
        button("reset") { @count = 0 },
        spacing: 8.0
      ),
      text_field(@name, placeholder: "your name") { |s| @name = s },
      text("hello, #{@name}"),
      spacing: 12.0,
      padding: 16.0
    )
  end
end

run(Counter.new, title: "counter")

File: rakugan/demo/counter.pl
# The reference: everything in this file is what Rakugan takes. The app
# is a class, its state is its fields, `view` is a method, and a handler
# is an anonymous sub that closes over the fields.
#
#   rakugan run  demo/counter.pl
#   rakugan gate demo/counter.pl --script "click:+1,input:Momo"
use Rakugan;

class Counter {
    use Rakugan;
    field $count = 0;
    field $name  = "";

    method view {
        column(
            text("count: $count", size => 34),
            row(
                button("+1",    on_click => sub { $count += 1 }),
                button("+10",   on_click => sub { $count += 10 }),
                button("reset", on_click => sub { $count = 0 }),
                spacing => 8,
            ),
            text_field($name, placeholder => "your name",
                       on_change => sub ($s) { $name = $s }),
            text("hello, $name"),
            spacing => 12,
            padding => 16,
        );
    }
}

run(Counter->new, title => "counter");
